jmclient/jmclient/storage.py

```diff
@@ -0,0 +1,333 @@
+        # interactive logins. For encryption we want something much slower.
+        self.settings = {
+            'time_cost': time_cost,
+            'memory_cost': memory_cost,
+            'parallelism': parallelism,
+            'hash_len': hash_len,
+            'type': argon2_type,
+            'version': version
+        }
+        self.salt = salt if salt is not None else get_random_bytes(salt_len)
+        self.hash = low_level.hash_secret_raw(password, self.salt,
+                                              **self.settings)
+
+
+class StorageError(Exception):
+    pass
+
+
+class StoragePasswordError(StorageError):
+    pass
+
+
+class Storage(object):
+    """
+    Responsible for reading/writing [encrypted] data to disk.
+
+    All data to be stored must be added to self.data which defaults to an
+    empty dict.
+
+    self.data must contain serializable data (dict, list, tuple, bytes, numbers)
+    Having str objects anywhere in self.data will lead to undefined behaviour (py3).
+    All dict keys must be bytes.
+
+    KDF: argon2, ENC: AES-256-CBC
+    """
+    MAGIC_UNENC = b'JMWALLET'
+    MAGIC_ENC =   b'JMENCWLT'
+    MAGIC_DETECT_ENC = b'JMWALLET'
+
+    ENC_KEY_BYTES = 32  # AES-256
+    SALT_LENGTH = 16
+
+    def __init__(self, path, password=None, create=False, read_only=False):
+        """
+        args:
+          path: file path to storage
+          password: bytes or None for unencrypted file
+          create: create file if it does not exist
+          read_only: do not change anything on the file system
+        """
+        self.path = path
+        self._lock_file = None
+        self._hash = None
+        self._data_checksum = None
+        self.data = None
+        self.changed = False
+        self.read_only = read_only
+        self.newly_created = False
+
+        if not os.path.isfile(path):
+            if create and not read_only:
+                self._create_new(password)
+                self._save_file()
+                self.newly_created = True
+            else:
+                raise StorageError("File not found.")
+        else:
+            self._load_file(password)
+
+        assert self.data is not None
+        assert self._data_checksum is not None
+
+        self._create_lock()
+
+    def is_encrypted(self):
+        return self._hash is not None
+
+    def is_locked(self):
+        return self._lock_file and os.path.exists(self._lock_file)
+
+    def was_changed(self):
+        """
+        return True if data differs from data on disk
+        """
+        return self._data_checksum != self._get_data_checksum()
+
+    def change_password(self, password):
+        if self.read_only:
+            raise StorageError("Cannot change password of read-only file.")
+        self._set_hash(password)
+        self._save_file()
+
+    def save(self):
+        """
+        Write file to disk if data was modified
+        """
+        if not self.was_changed():
```

Comment:
This may be a stupid comment but ... suppose the file got accidentally deleted since it was last persisted? `self.data` is in memory right, so it'll be seen as unchanged and won't get persisted.